packages/git/src/browser/blame/blame-contribution.ts

```diff
@@ -0,0 +1,140 @@
+                const editor = this.currentCodeEditor;
+                if (editor) {
+                    this.showBlame(editor);
+                }
+            },
+            isVisible: () =>
+                !!this.currentCodeEditor,
+            isEnabled: () => {
+                const editor = this.currentCodeEditor;
+                return !!editor && this.isBlameable(editor.editor.uri);
+            }
+        });
+        commands.registerCommand(BlameCommands.CLEAR_GIT_ANNOTATIONS, {
+            execute: () => {
+                const editor = this.currentCodeEditor;
+                if (editor) {
+                    this.clearBlame(editor.editor.uri);
+                }
+            },
+            isVisible: () =>
+                !!this.currentCodeEditor,
+            isEnabled: () => {
+                const editor = this.currentCodeEditor;
+                const enabled = !!editor && this.appliedDecorations.has(editor.editor.uri.toString());
+                return enabled;
+            }
+        });
+    }
+
+    protected get currentCodeEditor(): EditorWidget | undefined {
+        const editor = this.editorManager.currentEditor;
+        if (editor) {
+            if (!DiffUris.isDiffUri(editor.editor.uri)) {
+                return editor;
+            }
+        }
+        return undefined;
+    }
+
+    protected isBlameable(uri: string | URI): boolean {
+        return this.blameManager.isBlameable(uri.toString());
+    }
+
+    protected appliedDecorations = new Map<string, Disposable>();
+
+    protected async showBlame(editorWidget: EditorWidget) {
+        const uri = editorWidget.editor.uri;
+        let content = undefined;
+        if (editorWidget.editor.document.dirty) {
+            content = editorWidget.editor.document.getText();
+        }
+        const blame = await this.blameManager.getBlame(uri.toString(), content);
+        if (blame) {
+            const toDispose = new DisposableCollection();
+            this.appliedDecorations.set(uri.toString(), toDispose);
+            const editor = editorWidget.editor;
+            toDispose.push(this.decorator.decorate(blame, editor, editor.cursor.line));
+            toDispose.push(editor.onDocumentContentChanged(() => this.clearBlame(uri)));
+            toDispose.push(editor.onCursorPositionChanged(position => {
+                if (!toDispose.disposed) {
+                    this.decorator.decorate(blame, editor, position.line);
+                }
+            }));
+            editorWidget.disposed.connect(() => this.clearBlame(uri));
+        }
+    }
+
+    protected clearBlame(uri: string | URI) {
+        const decorations = this.appliedDecorations.get(uri.toString());
+        if (decorations) {
+            this.appliedDecorations.delete(uri.toString());
+            decorations.dispose();
+        }
+    }
+
+    registerMenus(menus: MenuModelRegistry): void {
+        menus.registerMenuAction([...EDITOR_CONTEXT_MENU, '3_git'], {
```

Comment:
it would be nice to introduce a constant for it in `GitViewContribution`.

Reply:
like this?

```
export namespace EditorContextMenu {
    export const GIT = [...EDITOR_CONTEXT_MENU, '3_git'];
}
```

Reply:
should be a different NS

Reply:
what about just a `export const EDITOR_CONTEXT_MENU_GIT = [...EDITOR_CONTEXT_MENU, '3_git']`?

Reply:
done.